tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_right_shift.h

```diff
@@ -0,0 +1,40 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include "ckernel.h"
+#include "ckernel_defs.h"
+#include "noc_nonblocking_api.h"
+
+using namespace sfpi;
+
+namespace ckernel {
+namespace sfpu {
+
+template <bool APPROXIMATION_MODE, int ITERATIONS = 8>
+inline void calculate_right_shift(uint value) {
+#pragma GCC unroll 0
+    for (int d = 0; d < ITERATIONS; d++) {
+        vUInt val = dst_reg[0];
+        vInt v = val;
+        v_if(val<0){
```

Comment:
How does the conditional check for a vUInt work here? Im assuming because its re-interpreting the vFloat bits from dst_reg[0] to vUInt, but I am unsure of the conditional check for vUint < 0 triggers, because it would assume its in the vuint range? 0 to 65536 ?

Reply:
If you remove this conditional check, does the PCC change?

Reply:
If we remove the conditional check it will fail, but as you said even though the vUInt range is 0 to 65536, if we pass a negative value, it should store as positive, but in this case it is still storing it as negative, thats why we have introduced the condition check for negative numbers
![Screenshot 2024-06-03 at 2 45 27 PM](https://github.com/tenstorrent/tt-metal/assets/156493059/40f0ee99-99b2-470f-8d72-839751761f6a)

Reply:
@pgkeller, How does the sfpi compile a v_if conditional check for a `vUint`? I would assume that this:
```
        vUInt val = dst_reg[0];
        v_if(val<0){
             <some code....>
        }
```
would always trigger false? Why is it needed in the above case?

Reply:
I just synced with @pgkeller, and he is also doubtful of why a conditional check for negatives would be needed for a VUint. I also see that you are testing inputs of signed integers?
```
datagen_func = [
            generation_funcs.gen_func_with_cast(partial(generation_funcs.gen_rand, low=-100, high=100), torch.int)
```
so I am unsure how you are reading signed integer values into a vUint at all?

Reply:
Ok I tested this locally, and it passes:

```
inline void calculate_right_shift(const uint shift_amt) {
#pragma GCC unroll 0
    for (int d = 0; d < ITERATIONS; d++) {
        vInt input = dst_reg[0];
        vUInt val = reinterpret<vUInt>(input);

        v_if(input<0){
            val = setsgn(val-1, 0);
        }
        v_endif;
        vInt res = reinterpret<vInt>(val >> shift_amt);

        v_if(input<0){
            res = setsgn(res+1, input);
        }
        v_endif;

        dst_reg[0] = res;
        dst_reg++;
    }
}
```

It makes the casting between vUint and Int explicit, and does the if check with a signed integer. @pgkeller will investigate later when he as cycles why the sfpi v_if check for negatives executes with vUint, might be a sfpi compilation issue.

Reply:
Sure @rtawfik01, We've updated the code.